feat: NFT Marketplace

green-jay opened the pull request:
## Description
NFT Marketplace

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+use primitives::ReserveIdentifier;
+
+mod benchmarking;
+pub mod types;
```
Does this need to be pub ?

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+#[cfg(test)]
+mod tests;
+
+pub type BalanceOf<T> =
+	<<T as pallet_nft::Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+pub type TokenInfoOf<T> = TokenInfo<BalanceOf<T>>;
+pub type OfferOf<T> =
```
to `pub` or not to `pub`?

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+	#[pallet::storage]
+	#[pallet::getter(fn tokens)]
+	/// Stores token info
+	pub type Tokens<T: Config> =
```
Same for the storage - i would not make pub accesible if not really needed. All storage types.

enthusiastmartin commented on primitives/Cargo.toml:
```diff
@@ -18,11 +18,11 @@ serde = { features = ["derive"], optional = true, version = "1.0.101" }
 frame-support = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.11", default-features = false }
 frame-system = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.11", default-features = false }
 sp-std = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.11", default-features = false }
+sp-runtime = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.11", default-features = false }
```
Why ?

green-jay replied:
needed for reserve identifier

enthusiastmartin commented on rococo-local/basilisk-karura.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"relaychain": {
-		"bin": "../../polkadot/target/release/polkadot",
+		"bin": "/mnt/data/workspace/sources/polkadot/target/release/polkadot",
```
this was my accidental change. revert all changes in this file.

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::OriginFor;
+
+	pub const RESERVE_ID: ReserveIdentifier = ReserveIdentifier::Marketplace;
```
also probably it does not need to be pub.

i would also add a brief comment what's the purpose of this.

PErhaps - you can make this as config option ?

green-jay replied:
removed pub and added a comment

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+		}
+
+		/// Set trading price and allow sell
+		/// Setting to NULL will disallow auto sell
```
```suggestion
		/// Setting price to None disables auto sell
```

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+				Error::<T>::NotTheTokenOwner
+			);
+
+			Tokens::<T>::try_mutate(class_id, instance_id, |maybe_token_info| -> DispatchResult {
+				let token_info = maybe_token_info.as_mut().ok_or(Error::<T>::NotListed)?;
+
+				token_info.price = new_price;
+
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::TokenPriceUpdated(sender, class_id, instance_id, new_price));
+
+			Ok(())
```
Can be simplified so you can get rid of one Ok.

```suggestion
			Tokens::<T>::try_mutate(class_id, instance_id, |maybe_token_info| -> DispatchResult {
				let token_info = maybe_token_info.as_mut().ok_or(Error::<T>::NotListed)?;

				token_info.price = new_price;
				
				Self::deposit_event(Event::TokenPriceUpdated(sender, class_id, instance_id, new_price));

				Ok(())
			})
```

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+
+		/// Lists the token on Marketplace
+		/// freezes the NFT from transfers
+		/// and other modifications
```
It would be better to describe the modifications as it might be confusing what it does.

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+				.map(|c| c.class_type)
+				.ok_or(pallet_nft::Error::<T>::ClassUnknown)?;
+
+			ensure!(
+				!Tokens::<T>::contains_key(class_id, instance_id),
+				Error::<T>::AlreadyListed
+			);
```
Perhaps move this before the first storage read which gets the class type.

I have a feeling it would be logically better ( so the class type read, check and validation is together )

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+			);
+			ensure!(class_type == ClassType::Marketplace, Error::<T>::UnsupportedClassType);
+
+			let royalty = Instances::<T>::get(class_id, instance_id)
+				.map(|i| i.royalty)
+				.ok_or(pallet_nft::Error::<T>::RoyaltyNotSet)?;
+			let author = Instances::<T>::get(class_id, instance_id)
+				.map(|i| i.author)
+				.ok_or(pallet_nft::Error::<T>::AuthorNotSet)?;
```
You read the same instance data twice. 

It would be better to get the data and then add something like:

```
ensure!( data.royalty, ... );
ensure!( data.author, ... );
```

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+
+			// Only token owner can list
+			ensure!(
+				pallet_uniques::Pallet::<T>::owner(class_id.into(), instance_id.into()) == Some(sender.clone()),
```
Would not be better to do this check as one of the first checks ? 

I feel lots of stuff is already read and checked prior to this.

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+
+		/// Unlists the token from Marketplace
+		/// unfreezes the NFT from transfers
+		/// and other modifications
```
Other modifications ... I'd prefer to mention them.

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+			Tokens::<T>::remove(class_id, instance_id);
+			Offers::<T>::remove_prefix(token_id, None);
+
+			let owner = pallet_uniques::Pallet::<T>::owner(class_id.into(), instance_id.into())
```
let's move this check before all storage changes.  

Even if it is transactional, it is better to check first and then make modifications.

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+		ClassOrInstanceUnknown,
+		/// Token already listed on marketplace
+		AlreadyListed,
+		/// Offer is lower than the pallet limit
```
```suggestion
		/// Offer is lower than the minimum limit
```

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+		UnknownOffer,
+		/// Offer is no longer valid
+		OfferExpired,
+		/// User already offered for this token
```
```suggestion
		/// User already made an offer for this token
```

enthusiastmartin reviewed (commented):
I haven't' got through it all yet. 

But here are my initial comments.

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+	/// Offer existence check by account and token id
+	#[pallet::storage]
+	#[pallet::getter(fn offers_by_maker)]
+	pub type OffersByMaker<T: Config> = StorageNMap<
```
I wonder why this cannot be combined into one storage together with the `Offers`.

If you make order of keys like `(NftClassId, NftInstanceId, AccountId)` and make the Offer type as storage item -> you actually get the same result .

green-jay replied:
true, initially i thought i would need remove offers by account, but i only need to remove offers by token id

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+		/// Offer was withdrawn \[sender, class_id, instance_id\]
+		OfferWithdrawn(T::AccountId, T::NftClassId, T::NftInstanceId),
+		/// Offer was accepted \[sender, class_id, instance_id\]
+		OfferAccepted(T::AccountId, T::NftClassId, T::NftInstanceId),
```
I would add also the amount of accepted offer.

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+		Tokens::<T>::try_mutate(class_id, instance_id, |maybe_token_info| -> DispatchResult {
+			let token_info = maybe_token_info.as_mut().ok_or(Error::<T>::NotListed)?;
+
+			let mut price;
+
+			if is_offer {
+				price = Offers::<T>::get(token_id, buyer.clone())
+					.map(|o| o.amount)
+					.ok_or(Error::<T>::UnknownOffer)?;
+			} else {
+				price = token_info.price.take().ok_or(Error::<T>::NotForSale)?;
+			}
```
```suggestion
			

			let mut price = if is_offer {
				Offers::<T>::get(token_id, buyer.clone())
					.map(|o| o.amount)
					.ok_or(Error::<T>::UnknownOffer)?
			} else {
			     token_info.price.take().ok_or(Error::<T>::NotForSale)?
			}
```

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+
+		let owner_origin = T::Origin::from(RawOrigin::Signed(owner.clone()));
+
+		pallet_uniques::Pallet::<T>::thaw(owner_origin, class_id.into(), instance_id.into())?;
```
Move this inside the try_mutate too. 

so in case try_mutates returns Err - this will be reverted too.

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+				price = token_info.price.take().ok_or(Error::<T>::NotForSale)?;
+			}
+
+			let royalty = Instances::<T>::get(class_id, instance_id)
+				.map(|i| i.royalty)
+				.ok_or(pallet_nft::Error::<T>::RoyaltyNotSet)?;
+			let author = Instances::<T>::get(class_id, instance_id)
+				.map(|i| i.author)
+				.ok_or(pallet_nft::Error::<T>::AuthorNotSet)?;
```
Same here as in previous case - get the data first and just ensure royalty and author.

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+				.ok_or(pallet_nft::Error::<T>::AuthorNotSet)?;
+
+			// Calculate royalty and subtract from price if author different from buyer
+			let royalty_perc = Percent::from_percent(royalty);
+			let royalty_amount = royalty_perc * price;
+
+			if owner != author && royalty != 0u8 {
+				price = price.saturating_sub(royalty_amount);
+
```
```suggestion


			if owner != author && royalty != 0u8 {
			       let royalty_perc = Percent::from_percent(royalty);
			        let royalty_amount = royalty_perc * price;
				price = price.saturating_sub(royalty_amount);

```

green-jay replied:
i need royalty info even if it's 0

enthusiastmartin commented on pallets/marketplace/src/lib.rs:
```diff
@@ -0,0 +1,486 @@
+			T::NftClassId,
+			T::NftInstanceId,
+			BalanceOf<T>,
+			Option<(T::AccountId, u8)>,
```
Why this needs to be option ? Currently as used - there is only one way.

enthusiastmartin reviewed (commented):
More comments on the marketplae pallets.

enthusiastmartin commented on pallets/marketplace/src/benchmarking.rs:
```diff
@@ -0,0 +1,136 @@
+		let caller = create_account::<T>("caller", 0);
+		let caller2 = create_account::<T>("caller2", 0);
+		pallet_nft::Pallet::<T>::create_class(RawOrigin::Signed(caller.clone()).into(), ClassType::Marketplace, bvec![0])?;
+		pallet_nft::Pallet::<T>::mint(RawOrigin::Signed(caller.clone()).into(), 0u32.into(), Some(caller.clone()), Some(20), Some(b"metadata".to_vec()))?;
```
Benchmarking : I believe this is metadata should be the longest possible to benchmark the worst case. 

I would just create the longest metadata and use it each benchmark where metadata is used

green-jay replied:
yes agreed, but it is possible that even some callable functions from pallet will be removed and some will appear so let's keep benchmarking as is and refactor as the last thing, will keep that on my list

enthusiastmartin commented on pallets/marketplace/src/benchmarking.rs:
```diff
@@ -0,0 +1,136 @@
+	accept_offer {
+		let caller = create_account::<T>("caller", 0);
+		let caller2 = create_account::<T>("caller2", 0);
+		pallet_nft::Pallet::<T>::create_class(RawOrigin::Signed(caller.clone()).into(), ClassType::Marketplace, bvec![0])?;
+		pallet_nft::Pallet::<T>::mint(RawOrigin::Signed(caller.clone()).into(), 0u32.into(), Some(caller.clone()), Some(20), Some(b"metadata".to_vec()))?;
+		Marketplace::<T>::list(RawOrigin::Signed(caller.clone()).into(), 0u32.into(), 0u32.into())?;
+		Marketplace::<T>::make_offer(RawOrigin::Signed(caller2.clone()).into(), 0u32.into(), 0u32.into(), unit(1_000).saturated_into(), 666u32.into())?;
```
just make this as an helper function - it is used in each benchmarks.

enthusiastmartin reviewed (commented):
few comments on marketplace benchmarks

enthusiastmartin reviewed (commented):
Few comments on NFT pallet.

green-jay commented on pallets/nft/src/lib.rs:
```diff
@@ -28,251 +31,303 @@ mod mock;
+	use frame_support::{pallet_prelude::*, traits::EnsureOrigin};
 	use frame_system::pallet_prelude::OriginFor;
 
+	pub const RESERVE_ID: ReserveIdentifier = ReserveIdentifier::Nft;
```
no I don't want this to be configurable, it's just an alias for better readability

green-jay commented on pallets/nft/src/lib.rs:
```diff
@@ -28,251 +31,303 @@ mod mock;
+	/// Next available class ID.
+	#[pallet::storage]
+	#[pallet::getter(fn next_class_id)]
+	pub type NextClassId<T: Config> = StorageValue<_, T::NftClassId, ValueQuery>;
```
storage will be accessed from other pallets

green-jay commented:
@enthusiastmartin I addressed your comments and upgraded to 0.9.12. pls check, I think we can merge this and start refactoring from there, pallet calls are disabled by filter

green-jay commented:
ok so in order to play it safe and avoid bugs in the node/runtime, let's not merge it until it's ready for live chain

vikiival commented:
> let's not merge it until it's ready for live chain

Maybe label `blocked` or `waiting for live` on the PR would be nice

martinfridrich reviewed (approved):
Full review is required but can be used for testing
